File: src/programmers/java/level2/게임_맵_최단거리.java
package programmers.level2;

import java.util.LinkedList;
import java.util.Queue;

public class 게임_맵_최단거리 {
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static int n, m;
    static int[][] dist;

    public static void main(String[] args) {
        int[][] maps = {{1, 0, 1, 1, 1}, {1, 0, 1, 0, 1}, {1, 0, 1, 1, 1}, {1, 1, 1, 0, 1}, {0, 0, 0, 0, 1}};
        solution(maps);
    }

    public static int solution(int[][] maps) {
        int answer;
        n = maps.length;
        m = maps[0].length;

        dist = new int[n][m];

        bfs(maps, dist, 0, 0);
        answer = dist[n - 1][m - 1];

        if (dist[n - 1][m - 1] == 0)
            answer = -1;

        return answer;
    }

    private static void bfs(int[][] maps, int[][] dist, int x, int y) {
        dist[x][y] = 1;
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{x, y});

        while (!queue.isEmpty()) {
            int[] current = queue.remove();
            int cx = current[0];
            int cy = current[1];

            for (int i = 0; i < 4; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if (nx >= 0 && ny >= 0 && nx < n && ny < m) {
                    if (dist[nx][ny] == 0 && maps[nx][ny] == 1) {
                        dist[nx][ny] = dist[cx][cy] + 1;
                        queue.add(new int[]{nx, ny});
                    }
                }
            }
        }
    }
}

File: src/programmers/javascript/level2/게임_맵_최단거리.js
function solution(maps) {
  const dx = [0, 0, 1, -1];
  const dy = [1, -1, 0, 0];

  const n = maps.length;
  const m = maps[0].length;

  let visited = Array.from(new Array(n), () => new Array(m).fill(false));

  const rangeCheck = (row, col) => {
    return row >= 0 && row < n && col >= 0 && col < m;
  };

  const BFS = (row, col) => {
    let queue = [[row, col]];
    visited[row][col] = true;

    while (queue.length !== 0) {
      let [x, y] = queue.shift();

      for (let i = 0; i < 4; i++) {
        let nx = x + dx[i];
        let ny = y + dy[i];

        if (rangeCheck(nx, ny) && maps[nx][ny] === 1 && !visited[nx][ny]) {
          visited[nx][ny] = true;
          maps[nx][ny] = maps[x][y] + 1;
          queue.push([nx, ny]);
        }
      }
    }
  };

  BFS(0, 0);

  return maps[n - 1][m - 1] > 1 ? maps[n - 1][m - 1] : -1;
}
